### ISSUE TITLE
Some false positives related to isset, is_null and assignment inside conditionals

### ISSUE THREAD

**Issue by kaan-atakan** [vimeo/psalm#2289]
I was prompted to create the following code ([psalm.dev](https://psalm.dev/r/5d02f87e8c)) to report a mistake with psalm that doesn't recognize that isset() also checks for null values:
```
<?php

  $var1 = rand(0,1) == 0 ? null : 'SomeClass';
  $var2 = null;
  $var4 = rand(0,1) == 0 ? null : 'SomeClass';
  $var5 = null;
  
  if(!isset($var0)) {
    $var0 = 'SomeClass';
  }

  if(!isset($var1)) {
    $var1 = 'SomeClass';
  }

  if(!isset($var2)) {
    $var2 = 'SomeClass';
  }

  if(!isset($var3) || is_null($var3)) {
    $var3 = 'SomeClass';
  }

  if(!isset($var4) || is_null($var4)) {
    $var4 = 'SomeClass';
  }

  if(!isset($var5) || is_null($var5)) {
    $var5 = 'SomeClass';
  }

  echo class_exists($var0) ? 1:0;
  echo class_exists($var1) ? 1:0;
  echo class_exists($var2) ? 1:0;
  echo class_exists($var3) ? 1:0;
  echo class_exists($var4) ? 1:0;
  echo class_exists($var5) ? 1:0;
```
Psalm produces the following errors and warnings:

```
ERROR: TypeDoesNotContainNull - 24:23 - Cannot resolve types for $var4 - string does not contain null
INFO: MixedArgument - 32:21 - Argument 1 of class_exists cannot be mixed, expecting string
ERROR: PossiblyNullArgument - 34:21 - Argument 1 of class_exists cannot be null, possibly null value provided
INFO: MixedArgument - 35:21 - Argument 1 of class_exists cannot be mixed, expecting string
INFO: MixedArgument - 37:21 - Argument 1 of class_exists cannot be mixed, expecting string
```

None of which make any sense to me. The MixedArgument errors are wrong because the variables will always be strings, not mixed.  The error on $var2 is what I originally wanted to report here, since isset also checks for null and psalm seems to not understand this. At least not always because I can't for the life of me understand how psalm then manages to correctly deduce at line 24:23 that $var4 is never null unless it bases it on the preceding !isset clause.

**Comment by muglug** [comment]
> The MixedArgument errors are wrong because the variables will always be strings, not mixed

Not for all of them - the types of `$var0` and `$var3` cannot be known definitively, and `$var5` is very unrealistic code - it's set to null, then there's a check whether it's set or null - that's weird enough that Psalm slightly panics.

Also, `class_exists` expects a class constant `SomeClass::class`.

This code is wonky enough that I don't think Psalm should treat it kindly - checking for class existence is reflection-level behaviour that shouldn't generally occur in regular business logic.
